### PR TITLE
Document implicit function try

### PR DESCRIPTION
**Pull request by lpil**
Inspired by this comment: https://github.com/elixir-lang/elixir/pull/5441#issuecomment-259980455 :)

### CODE REVIEW

**Comment by josevalim** [comment]
Awesome, thank you @lpil.

I am wondering if it should be documented under `Kernel.def` instead. This is not an ability added by `try` but by `def`. It is also supported in any other construct that cares to support it, such a `test` in ExUnit.

**Comment by lpil** [comment]
I was thinking this too. I'll move it over.

**Comment by fishcakez** [comment]
I do not like the wording "implicit try" because it implies TCO won't kick in and the function remains on the stack.

File: lib/elixir/lib/kernel.ex [hunk 1]
```diff
@@ -3334,6 +3334,25 @@ defmodule Kernel do
...
       #=> warning: the underscored variable "_bar" is used after being set
 
+  ## Implicit try
+
+  Function bodies serve as an implicit `try` block so `rescue` and `catch` can
+  be used without the `try` and `end`. The following two functions are
+  equivalent:
+
+      def format(value) do
+        try
```
Comment (fishcakez) [inline lib/elixir/lib/kernel.ex]: `try do`

**Comment by josevalim** [comment]
@fishcakez when moving to the docs for `def`, maybe we could call it "rescue/catch shortcut"?

File: lib/elixir/lib/kernel/special_forms.ex [hunk 2]
```diff
@@ -1816,7 +1816,6 @@ defmodule Kernel.SpecialForms do
         catch
           _, _ -> :failed
         end
-
```
Comment (lexmag) [inline lib/elixir/lib/kernel/special_forms.ex]: Looks like unnecessary change.

File: lib/elixir/lib/kernel.ex [hunk 3]
```diff
@@ -3334,6 +3334,25 @@ defmodule Kernel do
...
+          do_format(value)
```
Comment (lexmag) [inline lib/elixir/lib/kernel.ex]: Let's use `format!` instead of `do_format` here and below.

**Comment by fishcakez** [comment]
I am not sure of good wording :(

File: lib/elixir/lib/kernel.ex (hunk 1 above)
Comment (lpil) [inline lib/elixir/lib/kernel.ex]: Thanks!

**Comment by whatyouhide** [comment]
Good point @josevalim on adding it to `def` because it's added by `def` and not `try`, but I would still mention under `try` that "this construct is supported in a number of other places such as `Kernel.def`" (or something like this), just for discoverability. Wdyt?

**Comment by josevalim** [comment]
@whatyouhide I am not sure we want to make this feature super discoverable. 😅

**Comment by whatyouhide** [comment]
Well that is a good point :P

**Comment by lpil** [comment]
So do we have ideas for how we might want to document this? :)

File: lib/elixir/lib/kernel.ex [hunk 4]
```diff
@@ -3334,6 +3334,24 @@ defmodule Kernel do
...
+          :exit, reason -> {:error, reason}
+        end
+      end
+
+      def format(value) do
+        format!(value)
+      catch
+        :exit, reason -> {:error, reason}
+      end
   """
```
Comment (whatyouhide) [inline lib/elixir/lib/kernel.ex]: We usually leave an empty line between the last line of a code sample and the closing `"""` :)

Comment (lpil) [inline lib/elixir/lib/kernel.ex]: Thanks :)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by eksperimental** [comment]
thank you @lpil for the PR!

**Comment by lpil** [comment]
Thanks :)
